hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/AbstractJoiner.java

```diff
@@ -201,7 +194,15 @@ private void ensureConnectionToAllMembers() {
                         }
                     }
                 }
-            }
+                if (!allConnected) {
```

Comment:
We could do a 
```java
if (allConnected) {
    break;
}
```
here to ease the boolean logic of `sleep` and `while`. We could even keep the old `while-do` loop with it, which is IMO a bit easier to read.

Reply:
We could even put the `boolean allConnected` completely in the loop:
```java
            while (checkCount++ < connectAllWaitSeconds) {
                boolean allConnected = true;
                Collection<Member> members = clusterService.getMembers();
                for (Member member : members) {
                    if (!member.localMember() && node.connectionManager.getOrConnect(member.getAddress()) == null) {
                        allConnected = false;
                        if (logger.isFineEnabled()) {
                            logger.fine("Not-connected to " + member.getAddress());
                        }
                    }
                }
                if (allConnected) {
                    break;
                }
                try {
                    //noinspection BusyWait
                    TimeUnit.SECONDS.sleep(1);
                } catch (InterruptedException ignored) {
                    EmptyStatement.ignore(ignored);
                }
            }
```

Reply:
The `do-while`-> `while` looks good and seems better but it changes the behaviour in the pathological case when someone configures hazelcast with `GroupProperty#CONNECT_ALL_WAIT_SECONDS` set to 0. In this case, we don't print out to which members we are not connected to. Same thing with the `MockJoiner`, if we set `GroupProperty#MAX_JOIN_SECONDS` to 0, we wouldn't perform the logic in the loop at all.
@mdogan WDYT?